### ISSUE TITLE
manager.PackagesDownloadStarted not fired

### ISSUE THREAD

**Issue by KillTrot** [ProgTrade/nUpdate#29]
Hey guys,

Maybe i'm just not good enough with C#, but for me it seems that the event manager.PackagesDownloadStarded is not fired.

My Code:
```csharp
using nUpdate.Updating;
using System;
using System.Diagnostics;
using System.Globalization;
using System.Threading;
using System.Windows;

namespace KLauncher____Installer {
 /// <summary>
 /// Interaktionslogik für MainWindow.xaml
 /// </summary>
 public partial class MainWindow: Window {
  public MainWindow() {
   InitializeComponent();
  }

  private void button_Click(object sender, RoutedEventArgs e) {
   UpdateManager manager = new UpdateManager(new Uri("ServerURL"), "Key provide by nUpdate Administration", new CultureInfo("de-DE"));
   manager.PackagesDownloadStarted += Manager_PackagesDownloadStarted;
   var updaterUI = new UpdaterUI(manager, SynchronizationContext.Current);
   updaterUI.ShowUserInterface();
   manager.RestartHostApplication = false;

  }
  public void Manager_PackagesDownloadStarted(object sender, EventArgs e) {
   MessageBox.Show("Event fired!");
  }
 }
}
```

Please tell what i'm doing wrong:)

Have nice easter days.

**Comment by ProgTrade** [comment]
Hi,

there are a bunch of misunderstandings here with regards to the `UpdaterUI` class.
First: You must set your properties before starting the update process.
The line `manager.RestartHostApplication = false;` should consequently be placed above the call of `updaterUI.ShowUserInterface()`. The reason is that the class `UpdaterUI` does everything automatically in this method (seaching, downloading and installing) and thus, your program will be terminated at the end of this method, not taking care of your property. In any case, the `UpdateManager` instance must already be fully initialized when you're handling it over as argument to the constructor of `UpdaterUI` as this one keeps its instance in a backing field not noticing any further changes.

The second problem, that is also responsable for the event not being fired, can be split into two parts:
The `UpdaterUI` class, as already mentioned, implements the whole update process over the `UpdaterManager`, so that the user does not have to do anything else than calling `ShowUserInterface()`.
Now, there are two configurations of nUpdate. The one is `ProvideTAP` (offering the Taskbased asynchronous pattern) which you may be using, I think, and the other one is `WithoutTAP` (offering the Eventbased asynchronous pattern only). `WithoutTAP` will be removed in v4 btw.
Now, the `UpdaterUI` is determining the used configuration and either uses Tasks + async/await, or async methods with events then. But, in any case, even if you use EAP, the class is not designed to fire any events aiming that the user should handle them. The reason is, as already described, that everything is done automatically and there is no need for subscription of events. The events should only be used when you're implementing a custom UI or no UI at all. That would mean that you only use the `UpdateManager` itself and in this case you may implement the updating process using events.
tl;dr: `UpdaterUI` is a finished implementation of the `UpdateManager` for offering the user an integrated user interface and besides that, the user does not have to do anything else. You can also have a look at it on GitHub to see how this is working internally.

**Comment by KillTrot** [comment]
Much thanks for your answer!

My root problem is, that nUpdate does not close my application while installing, therefore the user have to do it manually. This is the reason why im trying to catch the event.
May you help me a little bit out.

**Comment by ProgTrade** [comment]
Ah, I see. You're using WPF, that's the problem. But, in any case, that wouldn't have been the necessary event. ;)
Version 3.x of nUpdate requires a small workaround for WPF as nUpdate, by default, calls `Application.Exit` which will not close the host application correctly. Create a new class and add this:

```csharp
using nUpdate.Updating;

public class WPFUpdateManager : UpdateManager
{
    public WPFUpdateManager(Uri updateConfigurationFileUri, string publicKey, CultureInfo languageCulture) : base(updateConfigurationFileUri, publicKey, languageCulture)
    { }
    public override void TerminateApplication()
    {
        Environment.Exit(0);
    }
}

Then use that class instead of `UpdateManager`. This should fix your problem.

**Comment by KillTrot** [comment]
This works!

Thanks alot!

**Comment by ProgTrade** [comment]
See also: #31
This issue is now fixed and thus, a workaround is no longer necessary.
